src/Symfony/Component/Console/Application.php

```diff
@@ -1000,37 +1000,61 @@ protected function doRunCommand(Command $command, InputInterface $input, OutputI
             }
         }
 
-        if ($this->signalsToDispatchEvent) {
-            $commandSignals = $command instanceof SignalableCommandInterface ? $command->getSubscribedSignals() : [];
-
-            if ($commandSignals || null !== $this->dispatcher) {
-                if (!$this->signalRegistry) {
-                    throw new RuntimeException('Unable to subscribe to signal events. Make sure that the `pcntl` extension is installed and that "pcntl_*" functions are not disabled by your php.ini\'s "disable_functions" directive.');
-                }
+        $commandSignals = $command instanceof SignalableCommandInterface ? $command->getSubscribedSignals() : [];
+        if ($commandSignals || null !== $this->dispatcher && $this->signalsToDispatchEvent) {
+            if (!$this->signalRegistry) {
+                throw new RuntimeException('Unable to subscribe to signal events. Make sure that the `pcntl` extension is installed and that "pcntl_*" functions are not disabled by your php.ini\'s "disable_functions" directive.');
+            }
 
-                if (Terminal::hasSttyAvailable()) {
-                    $sttyMode = shell_exec('stty -g');
+            if (Terminal::hasSttyAvailable()) {
+                $sttyMode = shell_exec('stty -g');
 
-                    foreach ([\SIGINT, \SIGTERM] as $signal) {
-                        $this->signalRegistry->register($signal, static function () use ($sttyMode) {
-                            shell_exec('stty '.$sttyMode);
-                        });
-                    }
+                foreach ([\SIGINT, \SIGTERM] as $signal) {
+                    $this->signalRegistry->register($signal, static fn () => shell_exec('stty '.$sttyMode));
                 }
             }
 
-            if (null !== $this->dispatcher) {
+            if ($this->dispatcher) {
+                // We register application signals, so that we can dispatch the event
                 foreach ($this->signalsToDispatchEvent as $signal) {
-                    $event = new ConsoleSignalEvent($command, $input, $output, $signal);
+                    $event = new ConsoleSignalEvent($command, $input, $output, $signal, 0);
 
-                    $this->signalRegistry->register($signal, function () use ($event) {
+                    $this->signalRegistry->register($signal, function ($signal) use ($event, $command, $commandSignals) {
                         $this->dispatcher->dispatch($event, ConsoleEvents::SIGNAL);
+                        $exitCode = $event->getExitCode();
+
+                        // If the command is signalable, we call the handleSignal() method
+                        if (\in_array($signal, $commandSignals, true)) {
+                            $command->handleSignal($signal);
+                            // BC layer for Symfony <5.3
+                            if (method_exists($command, 'getExitCode')) {
+                                $exitCode = $command->getExitCode($signal, $exitCode);
```

Comment:
Can we rely on `ConsoleEvents::TERMINATE` at this stage? If doing so there is an option to handle exit code there. So there will not be a BC (if dispatcher is available)

Reply:
I guess there is some confusion here. It's not about the `$event->getExitCode()`, but about `$command->getExitCode()`. (event vs command)

I introduced a new method in the `SignalableCommandInterface`, but with a BC layer. So at runtime, this method might not exist (yet).